File: LeetCode/0055.Jump Game/Solution.java
class Solution {
	public boolean canJump(int[] nums) {
		int goal = nums.length - 1;
        int maxStep = 0;
        for (int i = 0; i <= maxStep; i++) {
			maxStep = Math.max(maxStep, nums[i] + i);
            if (maxStep >= goal) {
                return true;
            }
        }
        return false;
	}
}

File: LeetCode/0055.Jump Game/solution.rs
pub struct Solution;

impl Solution {
	pub fn can_jump(nums: Vec<i32>) -> bool {
    	let goal = nums.len();
        let mut maxstep = 0;
        for i in 0..goal {
            if i <= maxstep {
                maxstep = maxstep.max(i + nums[i] as usize);
                if maxstep >= goal - 1 {
                    return true;
                }
            }
        }
        false
   }
}

#[test]
fn test1() {
    assert_eq!(Solution::can_jump(vec![2, 3, 1, 1, 4]), true);
}

#[test]
fn test2() {
  assert_eq!(Solution::can_jump(vec![3, 2, 1, 0, 4]), false);
}

#[test]
fn test3() {
  assert_eq!(Solution::can_jump(vec![1, 1, 1, 0]), true);
}
